### ISSUE TITLE
Why you have to use a regressor for the teacher model?

### ISSUE THREAD

**Issue by junleiz** [clovaai/overhaul-distillation#14]
Hello:
     Thanks for your inspiring work. I do not understand why you have to use a regressor for t/s model like resnet110 and resnet50? Since their dimensions for their distillation position are the same. Are these regressors necessary？

Thank you

**Comment by bhheo** [comment]
Hi

As you said, if the dimension is same, distillation can be done without regressor.

But, without regressor, the i-th channel of student should express the i-th channel of teacher.
However, since the student is initialized to a random value, in some cases it may be difficult to represent the teacher's i-th channel with the student's i-th channel.
This is especially true for resnet architectures where channel swapping is unlikely due to skip-connection.

Representing the teacher's i-th channel in another student's channel does not significantly impair the feature's similarity and make distillation easier.
Thus, even if the dimensions are the same, using a regressor can be beneficial for distillation.
But, as I said, it is not necessary, distillation might work without regressor.

Similarity with regressor is based on [FitNets](https://arxiv.org/abs/1412.6550), and [this paper (Similarity of Neural Network Representations Revisited)](https://arxiv.org/pdf/1905.00414.pdf) contains a detailed analysis about this similarity. This paper might help you to understand more about this similarity.

**Comment by junleiz** [comment]
> Hi
> 
> As you said, if the dimension is same, distillation can be done without regressor.
> 
> But, without regressor, the i-th channel of student should express the i-th channel of teacher.
> However, since the student is initialized to a random value, in some cases it may be difficult to represent the teacher's i-th channel with the student's i-th channel.
> This is especially true for resnet architectures where channel swapping is unlikely due to skip-connection.
> 
> Representing the teacher's i-th channel in another student's channel does not significantly impair the feature's similarity and make distillation easier.
> Thus, even if the dimensions are the same, using a regressor can be beneficial for distillation.
> But, as I said, it is not necessary, distillation might work without regressor.
> 
> Similarity with regressor is based on [FitNets](https://arxiv.org/abs/1412.6550), and [this paper (Similarity of Neural Network Representations Revisited)](https://arxiv.org/pdf/1905.00414.pdf) contains a detailed analysis about this similarity. This paper might help you to understand more about this similarity.

Thank you very much for your patience. I will read this paper. I tried your code in another way. If I train the student network just with the mse loss between the student and teacher feas, and the position is just like yours. The differences are that I comment on the regressor and I only use the mse as a loss.  I found the student network could learn sth (i.e. it could have a relative high acc on CIFAR100). But if I add the regressors to the loss, it will not work if I just use the mse loss

**Comment by bhheo** [comment]
In my experience, my code with mse loss also works well.
So, I recommend you to check following components

1. Whether parameters of regressor is included in optimizer. 
If you use regressor, it have to be trained with distillation loss.
So, you have to use optimizer (optim.sgd) with regressor parameters.
#13 is one of this case. 

2. Scale of mse loss. My code adopts summation over feature dimensions
But, default of MSE loss is mean over feature dimensions.
If you want to use same ```alpha``` parameter, you have to use MSE(reduction='sum')

I think one of those issues causes bad performance on your code.
If you code is still not working after fixing the two, I'm sorry but I don't know what makes distillation not working.
